### ISSUE TITLE
Feature Request: BASE_URL Support for Speedtest Tracker

### ISSUE THREAD

**Issue by soakes** [alexjustesen/speedtest-tracker#1563]
### Is your feature request related to a problem? Please describe.

Currently, Speedtest Tracker does not support a BASE_URL configuration option. This presents a problem when the app is deployed behind a reverse proxy with a path prefix (e.g., `/tracker`), as the app does not properly redirect URLs to reflect this setup.

### Describe the solution you'd like

I would like to see support for a BASE_URL configuration parameter in Speedtest Tracker. This feature would allow administrators to specify a base URL path (e.g., `/tracker`) where the app is hosted behind a reverse proxy. With this enhancement, all internal app URLs would correctly reflect the base URL, ensuring proper navigation and functionality without manual adjustments.

### Additional context

Currently, when Speedtest Tracker is accessed through a reverse proxy at a path like `/tracker`, the URLs generated by the app do not account for this prefix. This leads to broken links and incorrect redirections within the application. Adding a BASE_URL setting would mitigate these issues by dynamically adjusting internal URLs based on the configured base path (`/tracker` in this example).

This feature is crucial for seamless integration of Speedtest Tracker into environments where reverse proxies and URL prefixes are standard practices for security and organizational reasons.

**Comment by Berserkir-Wolf** [comment]
I second this one - it would be brilliant for those of us who expose things behind SSL proxies.

**Comment by alexjustesen** [comment]
You should be able to update your `APP_URL` to include the subdirectory postfix, something along the lines of `APP_URL: 'to-my-domain.tld/speedy'`

**Comment by soakes** [comment]
Thanks @alexjustesen will give this a try shortly and will update this message with the results.

**Comment by alexjustesen** [comment]
Dropping this here for later Alex...

https://gist.github.com/rafaucau/8259400871607bfb0826ba7f40b28474

**Comment by alnviana** [comment]
This comment may be useful: https://github.com/laravel/framework/issues/51210#issuecomment-2081386508

I tested it here, putting it right after the HTTPS check allows me to access the login screen by just going to http://fqdn/subfolder/, but it gives error 405 when I try to log in. :/

**Comment by mircsicz** [comment]
I (obvisously) won't be able to go through this gist tomorrow but plan to  do so in the coming days:
https://gist.github.com/luttje/689290ad45dc564cd2d3270b0bbac381

**Comment by alexjustesen** [comment]
According to https://laravel.com/docs/12.x/requests#configuring-trusted-hosts the middleware is setup to support the Caddy config in @mircsicz post above. I don't run Caddy in my lab so I have no way of testing this.

If this still isn't working please feel free to open a new issue and reference this one.

**Comment by soakes** [comment]
> Laravel was expected to work transparently with a reverse proxy. However, this application doesn't work properly because some resources point to the wrong way. (e.g. http://fqdn/livewire/livewire.min.js instead of http://fqdn:8081/speedtest-tracker/livewire/livewire.min.js)
> 
> Here's how I made it work in my environment, just in case anyone needs it. In my case, I use Nginx as a reverse proxy. To correct the incorrect requests, I set filters to rewrite the HTML, .js, .css and .json resources. Therefore, all these requests are analyzed and rewritten at runtime, so there is an unnecessary waste of resources. As in my case only I access this application, I don't think this is a problem, but YMMV.
> 
> ```
> # Modify response's host header
> proxy_redirect $proxy_x_forwarded_proto://$http_host/ $proxy_x_forwarded_proto://$http_host/speedtest-tracker/;
> 
> # Modify the content sent to the browser
> sub_filter_once off;
> sub_filter_types text/css text/javascript application/javascript application/json;
> sub_filter '/livewire/' '/speedtest-tracker/livewire/';
> sub_filter '/admin' '/speedtest-tracker/admin';
> sub_filter '\/admin' '\/speedtest-tracker\/admin';
> ```

Thanks for sharing your workaround and the detailed Nginx configuration—it’s definitely helpful for anyone who needs a quick fix. That said, I agree with your assessment that this approach is more of a hack than a proper solution. Relying on runtime rewrites to patch up resource paths isn’t ideal, especially as it can introduce unnecessary complexity and overhead, and may not scale well for broader deployments.

Ideally, the application itself should handle reverse proxy scenarios correctly, generating resource URLs relative to the configured base path. This would make it much easier to deploy behind a reverse proxy without needing custom rewrite rules or content filtering.

It would be great to see native support for reverse proxy setups added to the app, so these kinds of workarounds aren’t necessary. Thanks again for documenting your approach, but I hope this can be addressed properly in the application itself.

**Comment by alnviana** [comment]
> > Laravel was expected to work transparently with a reverse proxy. However, this application doesn't work properly because some resources point to the wrong way. (e.g. http://fqdn/livewire/livewire.min.js instead of http://fqdn:8081/speedtest-tracker/livewire/livewire.min.js)
> > Here's how I made it work in my environment, just in case anyone needs it. In my case, I use Nginx as a reverse proxy. To correct the incorrect requests, I set filters to rewrite the HTML, .js, .css and .json resources. Therefore, all these requests are analyzed and rewritten at runtime, so there is an unnecessary waste of resources. As in my case only I access this application, I don't think this is a problem, but YMMV.
> > ```
> > # Modify response's host header
> > proxy_redirect $proxy_x_forwarded_proto://$http_host/ $proxy_x_forwarded_proto://$http_host/speedtest-tracker/;
> > 
> > # Modify the content sent to the browser
> > sub_filter_once off;
> > sub_filter_types text/css text/javascript application/javascript application/json;
> > sub_filter '/livewire/' '/speedtest-tracker/livewire/';
> > sub_filter '/admin' '/speedtest-tracker/admin';
> > sub_filter '\/admin' '\/speedtest-tracker\/admin';
> > ```
> 
> Thanks for sharing your workaround and the detailed Nginx configuration—it’s definitely helpful for anyone who needs a quick fix. That said, I agree with your assessment that this approach is more of a hack than a proper solution. Relying on runtime rewrites to patch up resource paths isn’t ideal, especially as it can introduce unnecessary complexity and overhead, and may not scale well for broader deployments.
> 
> Ideally, the application itself should handle reverse proxy scenarios correctly, generating resource URLs relative to the configured base path. This would make it much easier to deploy behind a reverse proxy without needing custom rewrite rules or content filtering.
> 
> It would be great to see native support for reverse proxy setups added to the app, so these kinds of workarounds aren’t necessary. Thanks again for documenting your approach, but I hope this can be addressed properly in the application itself.

Indeed. I have other systems that works right out of the box, some others need to change some kind of "base_url" to work right.
It makes more sense to the own application to give the right path, instead of doing rewrite.
What I can say is:  It is strange that the Laravel works, but I have to rewrite the livewire paths. Maybe Laravel supports proxy, but livewire doesn't? Or maybe the way it was implemented?
But as I don't have a great knowledge, there is a chance that I didn't configure the proxy properly, or even the Laravel part. So I don't want to insist in something that could be wrong and make someone mad.

**Comment by siment** [comment]
Hey guys. As much as I appreciate the tips on how to make subdirectories work by manipulating proxy requests, I don't think this feature request should be closed. It would still be a nice feature to be able to configure `BASE_URL` - just as the original requester described.

On a relevant note: Not everyone has enabled - or _can_ enable - the Nginx module `ngx_http_sub_module` for this hack to work.

A more robust solution would be for the web application to construct correct requests based on the `BASE_URL` parameter.

Thanks for all the good work!
